c-deps/libroach/ccl/key_manager.cc

```diff
@@ -7,54 +7,156 @@
 //     https://github.com/cockroachdb/cockroach/blob/master/licenses/CCL.txt
 
 #include "key_manager.h"
-#include <cryptopp/modes.h>
-#include <google/protobuf/stubs/stringprintf.h>
 #include "../fmt.h"
+#include "../utils.h"
+#include "crypto_utils.h"
 
 static std::string kFilenamePlain = "plain";
 
-static rocksdb::Status KeyFromFile(rocksdb::Env* env, const std::string& path, EncryptionKey* key) {
+namespace KeyManagerUtils {
+
+rocksdb::Status KeyFromFile(rocksdb::Env* env, const std::string& path, enginepbccl::SecretKey* key) {
+  int64_t now;
+  auto status = env->GetCurrentTime(&now);
+  if (!status.ok()) {
+    return status;
+  }
+
+  auto info = key->mutable_info();
   if (path == kFilenamePlain) {
-    // plaintext placeholder.
-    key->type = PLAIN;
-    key->id = "";
-    key->key = "";
-    key->source = kFilenamePlain;
+    key->set_key("");
+    info->set_encryption_type(enginepb::Plaintext);
+    info->set_key_id(kPlainKeyID);
+    info->set_creation_time(now);
+    info->set_source(kFilenamePlain);
     return rocksdb::Status::OK();
   }
 
   // Real file, try to read it.
   std::string contents;
-  auto status = rocksdb::ReadFileToString(env, path, &contents);
+  status = rocksdb::ReadFileToString(env, path, &contents);
   if (!status.ok()) {
     return status;
   }
 
   // Check that the length is valid for AES.
   auto key_length = contents.size();
-  if (key_length != 16 && key_length != 24 && key_length != 32) {
+  switch (key_length) {
+  case 16:
+    info->set_encryption_type(enginepb::AES128_CTR);
+    break;
+  case 24:
+    info->set_encryption_type(enginepb::AES192_CTR);
+    break;
+  case 32:
+    info->set_encryption_type(enginepb::AES256_CTR);
+    break;
+  default:
     return rocksdb::Status::InvalidArgument(fmt::StringPrintf(
         "key in file %s is %llu bytes long, AES keys can be 16, 24, or 32 bytes", path.c_str(), key_length));
   }
 
   // Fill in the key.
-  key->type = AES;
-  key->id = KeyHash(contents);
-  key->key = contents;
-  key->source = path;
+  key->set_key(contents);
+  info->set_key_id(KeyHash(contents));
```

Comment:
this is iffy. The key IDs are meant to be stored in the clear in the file registry. Using plain hashes makes it too easy to do lookup table attacks against key hashes.
We have some options here, some simpler than others:
* use the first 256 bits of the hash for the ID, the number of needed bytes after that for the key. We would need to switch to SHA-512 (not much of an issue).
* use salts per key. We only have two key IDs to check against anyway.

For the data keys, we don't need to match duplicate keys so we can simply generate a random ID as well.